### PR TITLE
Removes context.compile_internal where easy on `numba/cpython/listobj.py`

### PR DESCRIPTION
**Pull request by guilhermeleobas**
- list.copy
- list.reverse
- list.count
- list.index
- list.remove
- operator.ne
- operator.lt
- operator.le
- operator.gt
- operator.ge

### CODE REVIEW

File: numba/cpython/listobj.py [hunk 1]
```diff
@@ -794,16 +795,20 @@ def list_eq(context, builder, sig, args):
 
     return builder.load(res)
 
-@lower_builtin(operator.ne, types.List, types.List)
-def list_ne(context, builder, sig, args):
+@overload(operator.ne)
+def impl_list_ne(a, b):
+    if not all([isinstance(typ, types.List) for typ in [a, b]]):
+        return
```
Comment (stuartarchibald) [inline numba/cpython/listobj.py:801]: Perhaps factor this out into a function `all_lists(*args)` or similar?

File: numba/cpython/listobj.py [hunk 2]
```diff
@@ -921,57 +935,49 @@ def list_extend(lst, iterable):
...
-@lower_builtin("list.index", types.List, types.Any,
-           types.Integer)
-def list_index(context, builder, sig, args):
 
-    def list_index_impl(lst, value, start):
-        n = len(lst)
-        if start < 0:
-            start += n
+@overload_method(types.List, "index")
+def list_index(lst, value, start=None, end=None):
```
Comment (stuartarchibald) [inline numba/cpython/listobj.py:940]: This isn't the signature for `list.index`? It's, e.g.:
```pycon
In [1]: list.index?
Signature: list.index(self, value, start=0, stop=9223372036854775807, /)
```
I suspect the `stop` is `max(intp)`.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @guilhermeleobas. The refactor in general looks good, but I think some more work needs doing on `list.index` to match the CPython implementation (as noted inline). Thanks again!

File: numba/cpython/listobj.py (hunk 2 above)
Comment (guilhermeleobas) [reply numba/cpython/listobj.py]: Done!

File: numba/cpython/listobj.py [hunk 3]
```diff
@@ -921,57 +940,35 @@ def list_extend(lst, iterable):
...
-@lower_builtin("list.index", types.List, types.Any)
-def list_index(context, builder, sig, args):
 
-    def list_index_impl(lst, value):
-        for i in range(len(lst)):
-            if lst[i] == value:
-                return i
-        # XXX references are leaked when raising
-        raise ValueError("value not in list")
+intp_max = np.iinfo(as_dtype(types.intp)).max
```
Comment (stuartarchibald) [inline numba/cpython/listobj.py:944]: If possible, the `cpython` implementations should not rely on or refer to NumPy. In this case I think the `intp_max` can be found via:
```
>>> from numba import types
>>> types.intp.maxval
9223372036854775807
```

File: numba/cpython/listobj.py [hunk 4]
```diff
@@ -921,57 +940,35 @@ def list_extend(lst, iterable):
...
 
-    return context.compile_internal(builder, list_index_impl, sig, args)
 
-@lower_builtin("list.index", types.List, types.Any,
-           types.Integer)
-def list_index(context, builder, sig, args):
+@overload_method(types.List, "index")
+def list_index(lst, value, start=0, end=intp_max):
```
Comment (stuartarchibald) [inline numba/cpython/listobj.py:948]:
```
In [1]: list.index?
Signature: list.index(self, value, start=0, stop=9223372036854775807, /)
```
The last kwarg is named `stop` not `end`?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates @guilhermeleobas, couple of minor things to look at else looks good.

File: numba/cpython/listobj.py [hunk 5]
```diff
@@ -17,6 +17,7 @@
 from numba.misc import quicksort
 from numba.cpython import slicing
 from numba import literal_unroll
+from numba.np.numpy_support import as_dtype
```
Comment (stuartarchibald) [inline numba/cpython/listobj.py:20]:
```suggestion
```
dead code?

File: numba/cpython/listobj.py [hunk 6]
```diff
@@ -921,41 +938,19 @@ def list_extend(lst, iterable):
...
-                start = 0
-        for i in range(start, len(lst)):
-            if lst[i] == value:
-                return i
-        # XXX references are leaked when raising
-        raise ValueError("value not in list")
+    if not isinstance(start, (int, types.Integer, types.Omitted)):
+        raise TypeError(f'arg "start" must be an Integer. Got {start}')
+    if not isinstance(stop, (int, types.Integer, types.Omitted)):
+        raise TypeError(f'arg "end" must be an Integer. Got {end}')
```
Comment (stuartarchibald) [inline numba/cpython/listobj.py:951]: Just noticed that this refers to `end`, which is undefined. That it's not been caught by testing suggests that it is untested. Perhaps add a test to hit these two branches?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates, I just gave this a final check and noticed that there's a potential bug in the error reporting (noted inline). Once fixed, should be good to merge. Thanks again.

**Pull request merged**